### ISSUE TITLE
Feature Request: Stroke Order

### ISSUE THREAD

**Issue by turtlemaster19** [davidsansome/tsurukame#253]
Some users like to see the stroke order of kanji. This request is for the ability to access stroke order diagrams when viewing kanji.

**Comment by EmmaINSA** [comment]
It would be greatly appreciated, I totally agree with you! WaniKani is a great tool for learning kanji meaning and pronunciation (as well as vocabulary), but if you want to learn how to write kanji, it does not provide anything except radical breakdown. I would love to see this feature added to Tsurukame!

**Comment by mzsanford** [comment]
I looked at where the data for those stroke diagrams comes from and I have a few ideas. The KaniVG stroke diagrams can be numbered, animated (fade in), or a strip like the user script. That leads to a few questions:

 1. Where would we want to add these? The detail pages for Kanji only or somewhere in the lessons/reviews? I feel like the lessons/reviews would be difficult to add due to the limited space on screen. That leads me to …
 2. Which type of display works best given the space from point 1? I created a direction-animated version of the files also (example: https://mzsanford.com/share/20594_animated.svg) which can fit a smaller space, just in case that seems better. It assumes the KanjiVG data is in stroke direction order which seems to be true.

If I can get an answer to those I can take a look at implementing this. I don't plan to use this feature so I have the least knowledge about the answers to these questions.

## Technical Considerations

If we use numbered- or strip diagrams I think we can convert those to PNGs and use a UIImageView. If we want animation we'll need to use a UIWebView to display the SVG, which is a little more work.

We should also consider how we store/access these. To get offline access they could be bundled in the app, as [On Demand Resources](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/On_Demand_Resources_Guide/index.html), or download via Zip. The Zip download is how custom fonts currently work. The On Demand Resources would probably download them when the feature was enabled, similar to fonts. It would just offload the storage to Apple. Storage in the app would increase download sizes quite a bit I imagine.

**Comment by bmironer** [comment]
It order to learn the order while reviewing it be great to have the kanji draw itself during lessons and reviews, when the kanji appears, or maybe when tapping it ? We'd have to find a way to make it work with the custom font swipe too.
But that's one of the feature I'd really really wish the app had.

**Comment by turtlemaster19** [comment]
I think in the detail pages for sure, but I never really thought about it being in reviews. Maybe it being in reviews could be a setting: it would probably be mutually exclusive with the random fonts.

Basically, I agree with the comment above.

P.S. Don't forget about custom symbol images: https://developer.apple.com/documentation/uikit/uiimage/creating_custom_symbol_images_for_your_app

**Comment by EmmaINSA** [comment]
I think one way of integrating the stroke order feature could be to access it from each kanji's detail page, e.g. in a field between the 'radicals' and 'meaning explanation' ones.

If so, it could either be displayed as a static thumbnail of the kanji, leading to a full-size display of the animated stroke order like the .svg file @mzsanford linked when tapped, or the 'stroke order' field could be a 'stroke order' text link leading to the animated display ? 
I guess the second option would be easier to implement and would use less memory.

**Comment by mzsanford** [comment]
In reviewing https://github.com/davidsansome/tsurukame/issues/303 I was reminded of this issue. I'm going to pause my work on this issue for the moment. Read on if you want more info…

I fear that we might be adding too much dictionary functionality to what is at its core a language learning app. All of these are helpful but every one of them is using up (volunteer) development and maintenance time when purpose built dictionary apps exist and cover these features better than we can.

I'm not going to continue immediate work on this feature at the moment. I won't close this issue as someone else may disagree or time may still allow the addition of things like this later.

**Comment by bmironer** [comment]
Agreed.
I'm not sure it would be that useful as an bit of extra info on the kanji page either.

However, in reviews, I believe it would allow to memorise the stroke order at the same time, 
better grasp the radical decomposition, and probably help with the whole kanji too.
It's obviously a big undertaking and I was pleasantly surprised it was even considered.

For the record, I've just reached lesson 45, but I'm barely able to write kanjis from the first 10 lessons properly. Better start copying lines I guess... :)

**Comment by turtlemaster19** [comment]
@bmironer @mzsanford I agree. On reconsidering, I would love to see it in reviews, but maybe isn't as necessary on the kanji page, since it would probably take up too much space...

**Comment by AzzyLuddy** [comment]
Would love this! I use Tsurukame on my iPad Pro in splitscreen to write while I do reviews and lessons. Sick of using the Jisho website lol.

**Comment by turtlemaster19** [comment]
@mzsanford If you get some time and want to, can you redirect your stroke order work to reviews as discussed above? That wouldn't be too dictionary-like, but a nice optional feature in reviews so you know what it looks like writing!
